vtkVmtk/Segmentation/vtkvmtkAnisotropicDiffusionImageFilter.cxx

```diff
@@ -0,0 +1,237 @@
+Language:  C++
+Date:      $Date: 2006/04/06 16:48:25 $
+Version:   $Revision: 1.1 $
+
+  Copyright (c) Luca Antiga, David Steinman. All rights reserved.
+  See LICENCE file for details.
+
+  Portions of this code are covered under the VTK copyright.
+  See VTKCopyright.txt or http://www.kitware.com/VTKCopyright.htm 
+  for details.
+
+  Portions of this code are covered under the ITK copyright.
+  See ITKCopyright.txt or http://www.itk.org/HTML/Copyright.htm
+  for details.
+
+     This software is distributed WITHOUT ANY WARRANTY; without even 
+     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
+     PURPOSE.  See the above copyright notices for more information.
+
+=========================================================================*/
+
+#include "vtkvmtkAnisotropicDiffusionImageFilter.h"
+#include "vtkObjectFactory.h"
+#include "vtkType.h"
+
+#include "vtkvmtkITKFilterUtilities.h"
+
+#include "itkGradientAnisotropicDiffusionImageFilter.h"
+
+vtkStandardNewMacro(vtkvmtkAnisotropicDiffusionImageFilter);
+
+vtkvmtkAnisotropicDiffusionImageFilter::vtkvmtkAnisotropicDiffusionImageFilter()
+{
+  this->Conductance = 1.0;
+  this->TimeStep = 0.0625;
+  this->NumberOfIterations = 5;
+}
+
+
+template< class TImage >
+int vtkvmtkAnisotropicDiffusionImageFilter::FilterImage( vtkImageData* input,
+                                                         vtkImageData* output,
+                                                         typename TImage::Pointer image )
+{
+  using ImageType = TImage;
+  vtkvmtkITKFilterUtilities::VTKToITKImage<ImageType>(input,image);
+
+  typedef itk::GradientAnisotropicDiffusionImageFilter< ImageType, ImageType > AnisotropicDiffusionFilterType;
+  typename AnisotropicDiffusionFilterType::Pointer anisotropicDiffusionFilter = AnisotropicDiffusionFilterType::New();
+  
+  anisotropicDiffusionFilter->SetInput(image);
+  anisotropicDiffusionFilter->SetTimeStep(this->TimeStep);
+  anisotropicDiffusionFilter->SetNumberOfIterations(this->NumberOfIterations);
+  anisotropicDiffusionFilter->SetConductanceParameter(this->Conductance);
+
+  anisotropicDiffusionFilter->Update();
+  vtkvmtkITKFilterUtilities::ITKToVTKImage<ImageType>(anisotropicDiffusionFilter->GetOutput(),output);
+
+  return EXIT_SUCCESS;
+}
+
+
+template< int VDimension >
+int vtkvmtkAnisotropicDiffusionImageFilter::FilterScalarImage(vtkImageData* input,
+                                                              vtkImageData* output,
+                                                              int componentType )
+{
+  switch( componentType )
+  {
+    default:
+    case VTK_UNSIGNED_CHAR:
+    {
+      using PixelType = unsigned char;
+      using ImageType = itk::Image< PixelType, VDimension >;
+      typename ImageType::Pointer image = ImageType::New();
+      if( FilterImage< ImageType >( input, output, image ) == EXIT_FAILURE )
+      {
+        return EXIT_FAILURE;
+      }
+      break;
+    }
+
+    case VTK_CHAR:
```

Comment:
Shouldn't a char type fail every time. Not sure that the processing algorithm would be able to handle chars

Reply:
This portion was from something similar in ITK. I happy to remove the char type. I would assume that in most cases that makes sense. I don't have a good idea of which types are or should be handled.

Reply:
Ok. Yeah, from the documentation (https://itk.org/Doxygen/html/classitk_1_1GradientAnisotropicDiffusionImageFilter.html) it seems that only scalar image types are supported. Would probably be best to just remove it to prevent confusion in the future.

Reply:
I was a bit confused it seems. I've been doing a bit of Java and got confused between the C++ chars and Java chars (which are essentially unicode characters). The VTK_CHAR is just an 8 bit value (integers between 0 and 255), which seems like it's fine for this application. 

My understanding is that a scalar type (used in this context) is any numeric value that VTK can handle. In any case, it's not an issue, sorry about the confusion!